build.bash

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env bash
+
+nproc_default=4
+
+script_name=$(basename "${0}")
+
+show_help() {
+    cat << EOF
+Usage: ./$script_name [OPTIONS]
+
+Build script wrapper around CMake. Supplied arguments that do not match the
+options below will be passed to CMake when generating the build system.
+
+Options:
+      --clean   Delete build directory before invoking CMake.
+      --mpi     Compile MPI executable.
+      --compiler <compiler>
+                Specify the compiler to use.
+  -j <jobs>     Specify the number of parallel jobs in the compilation. By
+                default this value is set to $nproc_default.
+  -h, --help    Show this screen.
+
+Enabling debug mode:
+
+  The release build is default. To enable debug mode, specify the CMake option
+  -DCMAKE_BUILD_TYPE=Debug when invoking $script_name.
+
+Enabling verbose output from Makefile builds:
+
+  To enable more verbose output from Makefile builds, specify the CMake option
+  -DCMAKE_VERBOSE_MAKEFILE=ON when invoking $script_name.
+
+EOF
+}
+
+cmake_args=(-DCMAKE_BUILD_TYPE=Release)
+
+# Argument parsing adapted and stolen from http://mywiki.wooledge.org/BashFAQ/035#Complex_nonstandard_add-on_utilities
+while [ $# -gt 0 ]; do
+    case $1 in
+        --clean)
+            rm -r build
+            ;;
+        --mpi)
+            mpi=1
+            cmake_args+=(-DCABLE_MPI="ON")
+            ;;
+        --compiler)
+            compiler=$2
+            shift
```

Comment:
The standard in *nix tools is that the short form of options have the option, a space and then the argument, i.e. `-c gnu`, and the long form is with equal sign `--compiler=gnu`. Like this one can also pass arguments with a minus sign such as `--undef=-9999`. This is different here and confused me.
Why are you not using the `getopts` functionality of `bash`?

Reply:
I did look into `getopt` and `getopts` when writing the build script. A problem I had with `getopts` was that long-named options were not supported. `getopt` on the other hand supports long-named options however I was discouraged from using it for other reasons (see [here](http://mywiki.wooledge.org/ComplexOptionParsing)). In the end I went with the approach suggested here: http://mywiki.wooledge.org/BashFAQ/035#Complex_nonstandard_add-on_utilities.

Reply:
The link to the approach you gave shows it is possible to have a mix of short and long options and to have long options as `--option value` or `--option=value`. How we implement it comes down to personal choice. I don't have a preference either way, but I agree with @mcuntz that a lot of command-line tools use the form `--option=value`. 

I wouldn't support both `--option value` and `--option=value` because it makes the arguments parsing quite a lot longer for very little added advantage.

Reply:
All good to have long options with space. Makes the script the easiest to read. I guess:
```
-n)
    CMAKE_BUILD_PARALLEL_LEVEL=${2}
    shift
    ;;
--ncpus=?*)
    CMAKE_BUILD_PARALLEL_LEVEL=${1#*=}
    ;;

```
is not for everyone.

A last attempt from my side ;-) : why not having always a short option as well, such as -n for --ncpus, -m for --mpi, etc.